sstables/sstable_set.cc

```diff
@@ -769,6 +770,393 @@ sstable_set_impl::create_single_key_sstable_reader(
+    // because we may close readers in the middle of the partition query.
+    // In forwarding mode this is always false.
+    bool _should_return_partition_end;
+
+    // If a single reader wins with other readers (i.e. returns a smaller fragment) multiple times in a row,
+    // the reader becomes a ``galloping reader'' (and is pointed to by _galloping_reader).
+    // In this galloping mode we stop doing heap operations using the _peeked_readers heap;
+    // instead, we keep peeking the _galloping_reader and compare the returned fragment's position directly
+    // with the fragment of the reader stored at the heap front (if any), hoping that the galloping reader
+    // will keep winning. If he wins, we don't put the fragment on the heap, but immediately return it.
+    // If he loses, we go back to normal operation.
+    reader_iterator _galloping_reader;
+
+    // Counts how many times a potential galloping reader candidate has won with other readers.
+    int _gallop_mode_hits = 0;
+
+    // Determines how many times a fragment should be taken from the same
+    // reader in order to enter gallop mode. Must be greater than one.
+    static constexpr int _gallop_mode_entering_threshold = 3;
+
+    bool in_gallop_mode() const {
+        return _gallop_mode_hits >= _gallop_mode_entering_threshold;
+    }
+
+    // Relevant only in forwarding mode (in non-forwarding mode _halted_readers is always empty).
+    // Remove all halted readers (i.e. readers which returned end-of-stream) that we know for sure
+    // won't return any more fragments. We use the sstables' max_position() metadata to make the decision.
+    // Precondition: every halted reader returned all fragments with positions <= `up_to`.
+    // See also the comment for `_halted_readers`.
+    void collect_exhausted_readers(position_in_partition_view up_to) {
+        while (!_halted_readers.empty()
+                && _cmp(up_to, _halted_readers.front()->max_position) >= 0) {
+            boost::range::pop_heap(_halted_readers, _halted_reader_cmp);
+            _all_readers.erase(_halted_readers.back());
+            _halted_readers.pop_back();
+        }
+    }
+
+    // Retrieve the next fragment from the reader pointed to by `it`.
+    // The function assumes that we're not in galloping mode, `it` is in `_unpeeked_readers`,
+    // and all fragments previously returned from the reader have already been returned by operator().
+    //
+    // The peeked reader is pushed onto the _peeked_readers heap.
+    future<> peek_reader(db::timeout_clock::time_point timeout, reader_iterator it) {
+        return it->reader.peek(timeout).then([this, timeout, it] (mutation_fragment* mf) {
+            if (mf) {
+                if (mf->is_partition_start()) {
+                    // We assume there are no partition tombstones.
+                    if (mf->as_partition_start().partition_tombstone()) {
+                        on_internal_error(cormlogger, "partition tombstone encountered");
```

Comment:
Please include the partition key, also make the error message more verbose. This will be a very cryptic message for users (why is partition tombstone suddenly evil?).